### ISSUE TITLE
Celery batches and CELERY_TASK_ALWAYS_EAGER

### ISSUE THREAD

**Issue by nijel** [percipient/celery-batches#8]
I'd like to use `CELERY_TASK_ALWAYS_EAGER` for simple worker less setup and testing (I know it's not best way, but certainly simplest to begin with), but that doesn't seem to play well with celery-batches.

Is there some way to make eager mode work with batches or it's simply no go and I always need a worker to make it work?

**Comment by nijel** [comment]
Ah I just figure it out - in eager mode, the task doesn't get requests arg, but directly the args. Probably worth documenting somewhere.

**Comment by clokep** [comment]
Yeah, it's probably a bug that this doesn't happen automatically. I never really looked deep into whether this could be handled well.

I think in our code when I test this I usually build some basic structures to help with this (this won't run, but hopefully will give you the idea:

```python
from collections import namedtuple

from celery_batches import Batches

# app is your celery app

@app.task(base=Batches, ignore_result=True, flush_every=100, flush_interval=60)
def my_task(requests):
    for req in requests:
        print(req.args, req.kwargs)

# We have to mock out the grouping of requests that Celery passes to this task
BatchRequest = namedtuple('SimpleRequest', ['args', 'kwargs'])

def test_my_task():
    requests = [
        BatchRequest(args=[1, 2, 3], kwargs={'a': 'a', 'b': 'b'}),
        BatchRequest(args=[4, 5, 6], kwargs={'c': 'c', 'd': 'd'}),
    ]
    my_task(requests)
```

Of course testing this requests essentially testing the side-effects or mocking something inside the task, but that's pretty standard IMO.

Let me know if this helps at all. If nothing else, we probably should document this, yeah. 👍 If you come up with a clever way to do all this, please feel free to contribute your ideas!

**Comment by nijel** [comment]
So far I use simple boilerplate to process parameters, but that's also not really a clean solution, but at least it works:

```python
@shared_task(base=Batches, flush_every=500, flush_interval=300, bind=True)
def delete_fulltext(self, *args):
    # In eager mode we directly get params
    if isinstance(args[0], int):
        ids = [args]
    else:
        ids = [request.args for request in args]
```

**Comment by clokep** [comment]
For actually fixing this we would likely need to override `apply` in `Batches`, but I think that's what the worker itself ends up running too, so we would need to only do it in eager mode.

**Comment by clokep** [comment]
Out of curiosity, are you hoping that by calling `apply()` (or using always eager) it runs a batch of size 1 or are you hoping to be able to call some API to send a whole batch in at once?

**Comment by nijel** [comment]
I'd expect batch of size 1 in the eager mode. But both approaches would be probably okay

2. listopadu 2018 3:14:09 SEČ, Patrick Cloke <notifications@github.com> napsal:
>Out of curiosity, are you hoping that by calling `apply()` (or using
>always eager) it runs a batch of size 1 or are you hoping to be able to
>call some API to send a whole batch in at once?

**Comment by scalen** [comment]
Sorry to revive such an old issue, but I'm finding this a complete blocker for end-to-end automated testing.  If anyone has a fix, could they share it?

If not - my two cents:
`apply()` only appears to be called when in eager mode, while the worker unpacks a rabbit message then calls the underlying function (see below, and the `__call__()` method of tasks). https://github.com/percipient/celery-batches/blob/00c3fc777726e12c54678d1edc3d55e78d2ee5da/celery_batches/__init__.py#L175
On the other hand, I'm seeing some very strange behaviour when trying to override then call through to `super().apply()`... Apparently, `self` in `def apply(self, ...):` is coming through as `None`??

Does anyone have any ideas?

**Comment by clokep** [comment]
@scalen There's a couple of work-arounds above, but neither of those is great, create manually `Request` instances (see https://github.com/percipient/celery-batches/issues/8#issuecomment-415406435) has worked well for me for unit testing the task call itself. It breaks pretty hard if you're trying to test a function that *calls* a task. I usually get around this by making the task call into a `Mock` and just asserting *how* it was called.

I'm not sure what an actual fix would look like (see https://github.com/percipient/celery-batches/issues/8#issuecomment-435362440). How would you expect this work?

The short version of how I handle testing of batch tasks:
1. Manually create a `Request` and send a list of these into the task to unit test the task.
2. When testing a function that calls a batch task, mock the task call and use `assert_called_with`.

**Comment by scalen** [comment]
@clokep #16 is basically what I was thinking (basically what you suggested in https://github.com/percipient/celery-batches/issues/8#issuecomment-415451261).

Unfortunately, having created this and tested it to be working locally, it doesn't appear to work when imported into an actual project... It still seems to suffer from the problem I struck out in my previous comment https://github.com/percipient/celery-batches/issues/8#issuecomment-564025826.

**Comment by clokep** [comment]
I believe this will be fixed by #16.

**Comment by clokep** [comment]
I just pushed 0.3 which should support this, please let me know if something doesn't work here still!

**Comment by scalen** [comment]
Will do 🙂 thanks for all this!

**Comment by clokep** [comment]
Thanks for digging into this and fixing it @scalen!
